[move experiments] Regulated Coin

damirka opened the pull request:
This PR introduces two things:

### RegulatedCoin module

A module that provides a common implementable interface for handling regulated balances. Custom implementation that needs to instantiate a RegulatedCoin can implement any number of methods from RegulatedCoin. Main methods are:
```rs
public fun borrow<T: drop>(_: T, coin: &RegulatedCoin<T>): &Balance<T>;
public fun borrow_mut<T: drop>(_: T, coin: &mut RegulatedCoin<T>): &mut Balance<T>;
public fun zero<T: drop>(_: T, owner: address, ctx: &mut TxContext): RegulatedCoin<T>;

public fun from_balance<T: drop>(_: T, balance: Balance<T>, owner: address, ctx: &mut TxContext): RegulatedCoin<T>;
public fun into_balance<T: drop>(_: T, coin: RegulatedCoin<T>): Balance<T>;
```

### An example: ABC coin

This implementation aims to showcase one of the potential uses for the `RegulatedCoin`. The idea behind the implementation is to mimic real-world bank functionality. In detail:

1. Empty balances can only be created by admin (eg user goes to the bank and passes a KYC; a new balance is created by an admin);
2. Minting is also handled by admin (say, I transfer some amount of fiat currency to an organization, and they mint this coin in Sui and transfer it to my address);
3. Transfers between RC accounts are regulated - they are protected by a ban list and can only be received by the specified recipient;
4. Stolen instances of RegulatedCoin are "blocked" - only an owner can access his RC balance; in case he is banned, he cannot cheat by transferring his balance to someone else.

To make RC interoperable with the rest of the Sui-verse, there's an option to "cash out" a currency from a regulated account; the "cashed out" amount is wrapped as `Coin` and can be freely used by any user. An opposite action is putting a "cashed out" `Coin` back into the regulated account; can only be performed with accounts that haven't been banned.

The reasoning behind having two states of the same `Balance` are: 

1. Minting and Burning can only be performed with regulated coins (if there's some interaction with the real world).
2. Transferrable Coin is mimicking "cash" - it can be withdrawn or put into a bank account but cannot be used to interact with the outside world.

sblackshear commented on sui_programmability/examples/regulated_coin/Move.toml:
```diff
@@ -0,0 +1,9 @@
+[package]
+name = "USDC"
```
```suggestion
name = "RegulatedCoin"
```

sblackshear commented on sui_programmability/examples/regulated_coin/sources/RegulatedCoin.move:
```diff
@@ -0,0 +1,225 @@
+    }
+}
+
+module Stable::Hack {
```
I am still trying to understand what this example is saying...

damirka replied:
Poor implementation of contract security and the ability to consume a RegulatedCoin can lead to spawning Coin from balances of Regulated one.

sblackshear commented on sui_programmability/examples/regulated_coin/sources/RegulatedCoin.move:
```diff
@@ -0,0 +1,225 @@
+        ctx: &mut TxContext
+    ) {
+        let to_fill = Balance::split(REST::borrow_mut(coin, ctx), value);
+        let coin = Coin::mint<STAKE>(value, &mut stake.treasury_cap, ctx);
```
Nit: I found the shadowing of the `coin` parameter confusing here--a different name (`stake_coin`?) might help.

damirka replied:
Got lost in variables and naming. Nice catch!

sblackshear commented on sui_programmability/examples/regulated_coin/sources/RegulatedCoin.move:
```diff
@@ -0,0 +1,225 @@
+        Sui::ID::delete(id);
+    }
+
+    // === Explicit "Not Implemented" part ===
```
Wondering if we need/want this. For a client, I assume it would be easiest find out some functionality isn't supported by a module by checking whether it has the function(s) in question. Seeing that the functions exist, but discovering that they abort when trying them out seems less nice IMO.

damirka replied:
Yeah! I just meant it as an interface-like thing, and left it just to explicitly show that this is intentional. I wouldn’t pay additional gas for that if I were publishing it on the chain.

sblackshear reviewed (commented):
Like this approach a lot!!

lxfind commented on sui_programmability/examples/regulated_coin/sources/RegulatedCoin.move:
```diff
@@ -0,0 +1,431 @@
+    struct RegulatedCoin<phantom T> has key {
+        id: VersionedID,
+        balance: Balance<T>,
+        owner: address
```
I don't fully understand the need for the `owner` field. Do you anticipate a case where the Sui owner of the coin is different from the `owner` field of the coin?

damirka replied:
Yes indeed. And also it opens up a possibility to create lists of banned addresses. It is one of the possible usages of a regulated coin.

lxfind replied:
To implement banned addresses, you can always just check the sender address?
I still don't know when you would want the sender to be different from the owner field.

damirka replied:
Not quite. I assume that it is possible to send the whole struct to another user so he can withdraw as the sender address differs. By adding this field we give developers an option to prevent a scenario like this.

tnowacki replied:
> Not quite. I assume that it is possible to send the whole struct to another user so he can withdraw as the sender address differs. By adding this field we give developers an option to prevent a scenario like this.

I think that is a thing we need to prevent. I think it would be better to prevent the transfer than to add a secondary idea of ownership in the system. (though that doesn't necessarily mean we don't need an owner field, not sure. Just feel that the transfer should be regulated too)

tnowacki replied:
But in general what is going on here feels reasonable for the way transfers work today. But I think changes to how restricted transfers might work in the future will change this dramatically.

tnowacki commented on sui_programmability/examples/regulated_coin/sources/RegulatedCoin.move:
```diff
@@ -0,0 +1,575 @@
+/// - has a denylist for addresses managed by the coin admins
+/// - has restricted transfers which can not be taken by anyone except the recipient
+module ABC::ABC {
+    use RC::RegulatedCoin::{Self as RC, RegulatedCoin as RC};
```
Super minor nit, maybe just `RCoin`? I dream we might use `Rc` for the same thing as "ref counted" in Rust. It could be a good generalization of what's going on for global storage values inside the VM. 
(Imagine `borrow_global<T>(...): Rc<T>` or something. Could be interesting)

damirka replied:
Agreed! Used RC to write less code during development. 🐞

damirka commented:
@sblackshear @lxfind I've updated PR description, so you see the logic right away. 

Also, added an option to convert `RegulatedCoin<ABC>` to `Coin<ABC>`. Now it is possible for any RC owner to take and transfer it as a Coin.

lxfind commented:
I think this is still a draft? Or is it ready for review?
I still don't quite see the point of the `onwner` field of the `RegulatedCoin`.

lxfind commented:
My understanding is that the `owner` field of the `RegulatedCoin` provides constraint in the following scenario:
I have a RegulatedCoin, and I have just submitted a transfer request of this coin to account X (at this point X is not banned yet). After I submitted this transfer request, before X had a chance to accept the transfer, suddenly X gets banned, and we want to make sure we catch this case and forbid X from being able to accept this transfer.

Does the above sound right? This seems a very niche case to me that I am not sure why we want to support. Every other case can be solved by just filtering based on the owner of the coin object stored on-chain.

lxfind reviewed (dismissed):
Passing back for attention

sblackshear reviewed (approved):
LGTM, thanks Damir!!
